### PR TITLE
Plane: properly restart aborted landing

### PR DESCRIPTION
**Pull request by magicrub**
purpose is explained in issue https://github.com/diydrones/ardupilot/issues/2252

pretty straight forward implementation. When switching away from AUTO during a plane land, decrement mission index so that when we resume AUTO we would start the glide approach from the top waypoint.

All underflow conditions that I can think or are taken care of in set_current_cmd() so they were not needed here.

### CODE REVIEW

**Comment by magicrub** [comment]
I flew with this today, worked great for my repeated lidar tests!

**Comment by jinchengde** [comment]
@magicrub may I suggest set the next waypoint to the MAV_CMD_DO_LAND_START?

**Comment by magicrub** [comment]
Good idea!

**Comment by magicrub** [comment]
@jinchengde your great idea has ben incorporated and I'll be flight testing it later today

**Comment by magicrub** [comment]
flew this today while doing my lidar testing. I did multiple landing aborts /w DO_LAND_START, works great.

**Comment by magicrub** [comment]
flight tested and ready for merge

**Comment by magicrub** [comment]
rebased to master

**Comment by magicrub** [comment]
I just realized that an RTL during a land may get goofed up by this, or it would goof this up. I'll have to check that. Although if I rebase (again) I'll get the recent merged patch of ignoring RTL during a land. Hmmm...

**Comment by magicrub** [comment]
rebased to master again. I'll test the RTL thing soon

**Comment by magicrub** [comment]
re-re-rebased on master and RTLing on landing. Works great. Please merge

**Comment by iskess** [comment]
So what is the new procedure for Go-Around now? Does this require switching out of Auto and back in again? How does this work with DO_CONTINUE_AND_CHANGE_ALT release in 3.2? There was a lot of discussion last year on Drones Discuss with Tridge and Michael Day about how the Go-Around should be executed. https://groups.google.com/forum/#!msg/drones-discuss/XmVM9v4kIeo/H2oW2LyS--wJ
I felt that going max throttle while on final approach should trigger the go-around and sequence the next waypoint to DO_LAND_START. Has there been any movement on this?

**Comment by magicrub** [comment]
I was not aware of those discussions, they sound worthwhile to follow up on. This fixes the bug where if you have to abort a landing, by changing out of auto, it will restart the landing from the planned landing slope and glide path instead of just resuming from where ever you are at the time you switch back to auto.

i would consider this a component to the solution. The "full-throttle" stuff is a trigger, but still the behavior of what "go around" means is still unimplemented. This is that implementation.

TODO for additional work: when altitude is low, force a full-throttle pitch-up maneuver much like a takeoff to gain some altitude so there's no funny turns which could be dangerous.

**Comment by magicrub** [comment]
BUMP 
re-rebased on master (4th time), ready for merge.

**Comment by iskess** [comment]
Tom,
To execute a G/A using this change, can the operator simply switch out of Auto and right back again quickly? This should command a climb to the altitude of the WP after DO_START_LAND. If that WP is placed straight ahead, then there won't be a turn. 
Is this a viable work around to trigger a G/A and stay in Auto, until you can get to building the throttle lever trigger?

**Comment by magicrub** [comment]
@iskess Yes, that is the point. This is an RC controller-only solution so it's not very smart. With the addition of @mday299 's PR https://github.com/diydrones/ardupilot/pull/2715 I could add a feature where it would execute NAV_CONTINUE_AND_CHANGE_ALT if it is found after the LAND waypoint. That would be a separate PR, I'd like to get this one in as-is to at least get the mechanism in place and then we can augment from there..

**Comment by magicrub** [comment]
@iskess 
"Is this a viable work around to trigger a G/A and stay in Auto, until you can get to building the throttle lever trigger?"
This could be done too. The trigger can be anything we want, the hard part is to figure out how to act on it. Adding full-ish-throttle as an abort trigger would be a good PR to do after this goes through.

Speaking of going through.. *cough* @tridge *cough* please review before this turns into a monster feature-creep like is_flying did!

**Comment by mday299** [comment]
+1.  Good to clean up state when aborting landing.

**Comment by iskess** [comment]
So based on our discussions, the Go-Around development sequence can be described with the following steps:

1) With mode change sequence WP to DO_LAND_START (this PR)

2) If CONTINUE_AND_CHANGE_ALT is present, sequence to that first. 

3) Use full throttle stick as a trigger for Go-Around, using elements of the Take-off mode for pitch and power, but tracking the extended centerline of the approach path. The top is defined by either CONTINUE_AND_CHANGE_ALT, or the WP after DO_LAND_START, or the altitude of the final WP before LAND.

4) Modify the LAND command by adding an optional parameter in an empty field to define the G/A altitude. 

1st we need this PR to be reviewed.

**Comment by magicrub** [comment]
Thanks for the summary @iskess , I totally agree with that development order. Since this one is taking so long, and @mday299 's #2 is already in master, I'm tempted to integrate them together now. Naturally, I'd rather do it in steps but 3 months is getting a bit ridiculous. I've had to rebase this on master 4 times just to keep it relevant. I'd do it a 5th time ot integrate @mday299 's feature.

**Comment by magicrub** [comment]
I know @tridge will request this all to be documented on the wiki so lets jump on that

**Comment by tridge** [comment]
I'm concerned that this doesn't cope with jumps in missions. Just subtracting one from the WP number won't work in all cases - the previous WP in sequence order could be completely unrelated to the landing.
I like the idea of looking for the do_land_start.
Perhaps AP_Mission should keep the previous WP number as a state variable for this sort of use?

**Comment by magicrub** [comment]
I actually did implement the previous cmd thing because of the concerns you
brought up but eventually thought that no one would actually do that...
Probably a horrible assumption on my part.

There must be a nav waypoint before a land for a property glide slope but
there could be a do cmd in between. I could simply decrement the index
until it comes across a nav command that contains a waypoint.

**Comment by tridge** [comment]
we could also have gotten to the land WP from a jump. Going backwards in the WP list is never guaranteed to do the right thing.
I think we either need to remember the previous WP index in the vehicle code, or add an API in AP_Mission to remember the last nav WP. Doing it in the library is probably preferable as it can clear the validity of the previous WP when someone modifies the mission in flight

**Comment by magicrub** [comment]
K

**Comment by magicrub** [comment]
@Tridge,

I just noticed there is *mission.get_prev_nav_cmd_index()*. That is not
aware of nav cmds that contain waypoints though such at takeoff. I'mm make
 get_prev_nav_cmd_with_waypoint_index()

**Comment by magicrub** [comment]
Ok, all done tweaking this. Ready to merge!

**Comment by iskess** [comment]
Did you by chance get a chance to integrate it with CONTINUE_AND_CHANGE_ALT if it is present?

**Comment by magicrub** [comment]
I was having trouble with getting that working so it's left out for now.
I'll add it after this gets pushed in.

**Comment by magicrub** [comment]
@iskess btw, I'm working on it on a new branch - don't worry!

**Comment by iskess** [comment]
Great news, thank you Tom!

**Comment by magicrub** [comment]
@iskess Have you used CONTINUE_AND_CHANGE_ALT? Having trouble using it as it's designed.

**Comment by iskess** [comment]
I tried in in HITL and it worked. I had to use the MP actions tab to proceed direct to that WP.

**Comment by magicrub** [comment]
What do you mean? You couldn't switch modes?

**Comment by magicrub** [comment]
I got it working, it requires a waypoint before and a lat/lng assigned to it's own point. You also *must* go from a waypoint into that one, you can not be executing it then jump to, say, loiter, then back to auto. I fixed that in my _behavior PR

**Comment by magicrub** [comment]
Just flew PR https://github.com/diydrones/ardupilot/pull/2747 which is a branch based on this PR. Works great!

File: libraries/AP_Mission/AP_Mission.cpp [hunk 1]
```diff
@@ -372,6 +374,10 @@ bool AP_Mission::set_current_cmd(uint16_t index)
         if (is_nav_cmd(cmd)) {
             // save previous nav command index
             _prev_nav_cmd_index = _nav_cmd.index;
+            // save separate previous nav command index if it contains lat,long,alt
+            if (cmd.content.location.lat != 0 && cmd.content.location.lng != 0) {
```
Comment (tridge) [inline libraries/AP_Mission/AP_Mission.cpp]: i think this should be an || not an && (unless you have some prejudice against people landing on the equator?)

File: libraries/AP_Mission/AP_Mission.cpp [hunk 2]
```diff
@@ -1125,6 +1131,10 @@ bool AP_Mission::advance_current_nav_cmd()
...
```
Comment (tridge) [inline libraries/AP_Mission/AP_Mission.cpp]: same here

**Comment by tridge** [comment]
i've been testing this in SITL and while it does work as designed, I think we need some protection against the land restart triggering if the plane has already touched down, or the airspeed is below the stall speed.
The scenario I'm thinking of is that the plane does an auto-land, but runs long off the end of the runway into the grass (that happens to me with larger planes quite a lot). It is then sitting in the grass with poor RC connection to the pilot, and it could easily get a mode change. With this patch it will try to fly to the DO_LAND_START waypoint. The aircraft trying to power up and fly back will not go well.
If the plane is still on LAND_APPROACH I think this change is fine. If it has started the flare then it is only fine if the plane has enough airspeed to do a normal AUTO WP (note that it won't be doing a takeoff). So my preference would be to just make this auto-restart only happen on land approach

**Comment by tridge** [comment]
hmm, even on land approach this could be bad .... we'll need to think about it some more.
Imagine you have your landing airspeed set to 12m/s, but your normal auto cruise speed set to 20m/s. While doing an auto landing approach imagine you are at 5m altitude and you use this feature to switch out of auto then back into auto, expecting the plane to go back to the start of the landing.
What will happen is the plane will dive into the ground hard, with full throttle. When it re-enters auto it will be well below the target airspeed of 20m/s, so it wants to gain speed fast. It will do that by diving. It will also put on full throttle, and at the same time start doing a sharp turn. The airframe is unlikely to survive.
I tried this in SITL btw, and it does dive and turn at full throttle when you try this.

**Comment by magicrub** [comment]
I see your point, that *is* a major problem that I did not consider. The
other two PRs cascade on this will all suffer from that too.

How about we add a 5th flight stage to plane, one that is after land_final
perhaps called FLIGHT_LAND_ABORT = 5.

- keeps the current GO_AROUND behavior (allows NAV_LAND to complete)
- allows us to disable stall_prevention type behavior while in that stage
(no pitching down)
- allows the ability to do fancy things such as execute a hard-coded
CONTINUE_ALT or TAKEOFF command to pitch up and throttle up and after that
then we either jump to DO_LAND_START or previous good waypoint. No abort
planning needed!


-TomP

**Comment by iskess** [comment]
Just throwing this out there...please poke holes in it. 
What if we changed the current TAKEOFF command to TOGA. That is an airline industry term for Takeoff/Go-Around. The user can place a TOGA command as the first in a flight plan to act as the current TAKEOFF command, and/or he can place it just after the LAND command to define the Go-Around behavior (min pitch and altitude). If at anytime during Approach or Flare the throttle stick exceeds 95%, the plane will pitch and throttle up as defined by TOGA until reaching the chosen altitide. Then proceed to DO_LAND_START if it exists, else the last WP before LAND. 
Tom, I know you prefer a solution that requires less flight planning. I'm just not sure how to define the pitch angle without a flightplan command unless we turn that into another parameter. The G/A altitude would need to be defined by another command anyway, like CONTINUE_AND CHANGE_ALT or "TOGA". Unless you incorporate your idea of defining it with one of the LAND command's parameters. But you still need to define the pitch.

**Comment by magicrub** [comment]
Let's get some vocabulary sorted out. We have three things were talking
about adding:

- land abort trigger method (rc mode change, rc 95% throttle, MAVLink go
around command)
-abort behavior (pitch and throttle up, hold heading until altitude reached)
-go around (jump to do land start, decrement wp index, increment wp index
if continue_alt is present, increment index when go around issued by
MAVLink)

Here are some abort behavior options:
1. Put pitch and altitude in LAND mission content
2. use takeoff mission content and take off throttle param
3. create multiple LAND_GA_? params

My vote is to check for 1, then check for 2, then make a best guess at some
conservative numbers like 10deg and 50m.

**Comment by iskess** [comment]
Thanks for the definitions, I am carrying old baggage and using the term Go-Around to mean vertical path rather than abort. 
I like your 3 layered approach to this, it allows both flexibility and ease of use. 
It is tough to pick a good pitch angle to cover all planes without airspeed sensors. But since they have 2 other opportunities to define their pitch angle, I think it's very fair.

**Comment by magicrub** [comment]
If possible, I'd like to avoid using a param for behavior alt and angle and
throttle. I think it's best to just take it all from takeoff.

**Comment by magicrub** [comment]
And when I say use takeoff, I mean:

- Scan mission list for takeoff command, use first one it finds.
- Use it's altitude and pitch.
- execute in as much takeoff-ness as possible (flaps and such?)

I think having FLIGHT_LAND_ABORT=5 stage is the way to go, lets us do everything we want to do much easier.

**Comment by iskess** [comment]
Can we still have the goal of modifying TAKEOFF to track the extended centerline of the approach path? Possibly using the method that CRUISE uses track to a fictitious WP? Grabbing the heading at the moment ABORT is triggered may not be ideal. Besides the effects of wind shear common at treetop hight, it is possible the pilot was attempting to nudge around an obstacle and may have altered the heading. 
I hope you still allow the option to use an ABORT altitude different than TAKEOFF. I liked your idea of putting it in the LAND command. I will be wanting to use a much lower altitude than I use for TAKEOFF so I can build a missed approach path around obstacles. I will be claiming only 10m before needing to turn out of the way of trees or powerlines. I takeoff from a location that has a clear path and that will be set to 50m.

**Comment by magicrub** [comment]
I was having to do that heading-lock workaround because of the mode change.
If we stay in auto during the abort behavior then we'll have more info
available, such as holding the LAND wp heading. The head-hold was a
workaround from switching between modes.

**Comment by magicrub** [comment]
Yes, I plan to do the 3 staged thing where it checks for these values in
this order:
- LAND mission content
- TAKEOFF mission content
- other hard-coded conservative "content" such as 10deg and 50m. Not ideal
but should safely should work for everyone all the time.

**Comment by magicrub** [comment]
After talking to tridge, he's happy with this particular PR as long as we don't abort while flaring. I agree with that. baby steps!

**Comment by magicrub** [comment]
@tridge I have confirmed that if you switch modes after a flare then the throttle unsuppresses and you can fly out of it. Try it for yourself on the current master! set LAND_FLARE_ALT to 50 and then once you drop under 50m then switch to loiter mode.

I fixed the lat/lng==0 thing, this is now ready as-is. Please merge!

**Comment by tridge** [comment]
ok, I'm convinced, merged!

**Pull request closed without merging**